tokio/src/task/scope/cancellation_token.rs

```diff
@@ -0,0 +1,255 @@
+    /// This function is only safe as long as the `wait_node`s address is guaranteed
+    /// to be stable until it gets removed from the queue.
+    unsafe fn try_wait(
+        &mut self,
+        wait_node: &mut ListNode<WaitQueueEntry>,
+        cx: &mut Context<'_>,
+    ) -> Poll<()> {
+        match wait_node.state {
+            PollState::New => {
+                if self.is_cancelled {
+                    // The event is already signaled
+                    wait_node.state = PollState::Done;
+                    Poll::Ready(())
+                } else {
+                    // Added the task to the wait queue
+                    wait_node.task = Some(cx.waker().clone());
+                    wait_node.state = PollState::Waiting;
+                    self.waiters.add_front(wait_node);
+                    Poll::Pending
+                }
+            }
+            PollState::Waiting => {
+                // The WaitForCancellationFuture is already in the queue.
+                // The event can't have been set, since this would change the
+                // waitstate inside the mutex.
+                // We don't need to update the Waker here, since we assume the
+                // Future is only ever polled from the same task. If this wouldn't
+                // hold, true, this wouldn't be safe.
+                Poll::Pending
+            }
+            PollState::Done => {
+                // We have been woken up by the event.
+                // This does not guarantee that the event is still set. It could
+                // have been reset it in the meantime.
+                Poll::Ready(())
+            }
+        }
+    }
+
+    fn remove_waiter(&mut self, wait_node: &mut ListNode<WaitQueueEntry>) {
+        // WaitForCancellationFuture only needs to get removed if it has been added to
+        // the wait queue of the Event. This has happened in the PollState::Waiting case.
+        if let PollState::Waiting = wait_node.state {
+            if !unsafe { self.waiters.remove(wait_node) } {
+                // Panic if the address isn't found. This can only happen if the contract was
+                // violated, e.g. the WaitQueueEntry got moved after the initial poll.
+                panic!("Future could not be removed from wait queue");
+            }
+            wait_node.state = PollState::Done;
+        }
+    }
+}
+
+/// A synchronization primitive which can be either in the set or reset state.
+///
+/// Tasks can wait for the event to get set by obtaining a Future via `wait`.
+/// This Future will get fulfilled when the event has been set.
+pub(crate) struct CancellationToken {
+    inner: Mutex<CancellationTokenState>,
+}
+
+// The Event is can be sent to other threads as long as it's not borrowed
+unsafe impl Send for CancellationToken {}
+// The Event is thread-safe as long as the utilized Mutex is thread-safe
+unsafe impl Sync for CancellationToken {}
```

Comment:
I'm assuming this is inherited from `futures-intrusive` — AFAICT, we should automatically be `Send` + `Sync` without these impls, since we are not generic over a mutex?

Reply:
Afaik the issue was the raw pointer for the linked waiter list which is stored inside the struct. As long this is there the struct can't be `Send`, and the Mutex can't make it sync.
I can check if those are still necessary, but I guess so.